win32ss/user/winsrv/consrv/frontends/terminal.c

```diff
@@ -607,16 +607,56 @@ ConioWriteConsole(PFRONTEND FrontEnd,
         UpdateRect.Left  = min(UpdateRect.Left, Buff->CursorPosition.X);
         UpdateRect.Right = max(UpdateRect.Right, Buff->CursorPosition.X);
 
-        Ptr = ConioCoordToPointer(Buff, Buff->CursorPosition.X, Buff->CursorPosition.Y);
-
+        /* For Chinese, Japanese and Korean: */
         if (bCJK && Buffer[i] >= 0x80 && mk_wcwidth_cjk(Buffer[i]) == 2)
         {
             /* Buffer[i] is a fullwidth character */
-            /* FIXME */
-        }
 
-        Ptr->Char.UnicodeChar = Buffer[i];
-        if (Attrib) Ptr->Attributes = Buff->ScreenDefaultAttrib;
+            if (Buff->CursorPosition.X > 0)
+            {
+                /* Kill the previous leading byte */
+                Ptr = ConioCoordToPointer(Buff, Buff->CursorPosition.X - 1, Buff->CursorPosition.Y);
+                if (Ptr->Attributes & COMMON_LVB_LEADING_BYTE)
+                {
+                    Ptr->Char.UnicodeChar = L' ';
+                    if (Attrib)
+                        Ptr->Attributes &= ~COMMON_LVB_LEADING_BYTE;
+                }
+            }
+
+            if (Buff->CursorPosition.X == Buff->ScreenBufferSize.X - 1)
+            {
+                /* New line */
+                if (Buff->Mode & ENABLE_WRAP_AT_EOL_OUTPUT)
+                {
+                    Buff->CursorPosition.X = 0;
+                    ConioNextLine(Buff, &UpdateRect, &ScrolledLines);
+                }
+                else
+                {
+                    Buff->CursorPosition.X = CursorStartX;
+                }
+            }
+
+            /* Set leading */
+            Ptr = ConioCoordToPointer(Buff, Buff->CursorPosition.X, Buff->CursorPosition.Y);
+            Ptr->Char.UnicodeChar = Buffer[i];
+            if (Attrib)
+                Ptr->Attributes = Buff->ScreenDefaultAttrib | COMMON_LVB_LEADING_BYTE;
+
+            /* Set trailing */
+            Buff->CursorPosition.X++;
+            Ptr = ConioCoordToPointer(Buff, Buff->CursorPosition.X, Buff->CursorPosition.Y);
+            if (Attrib)
+                Ptr->Attributes = Buff->ScreenDefaultAttrib | COMMON_LVB_TRAILING_BYTE;
+        }
+        else
+        {
+            Ptr = ConioCoordToPointer(Buff, Buff->CursorPosition.X, Buff->CursorPosition.Y);
+            Ptr->Char.UnicodeChar = Buffer[i];
+            if (Attrib)
+                Ptr->Attributes = Buff->ScreenDefaultAttrib;
+        }
 
         Buff->CursorPosition.X++;
```

Comment:
Since the code block that is below has been copied for the CJK case I was wondering whether it is OK to keep it there or whether it should go inside the "else" block above?